Brokerages/Binance/BinanceBrokerage.cs

```diff
@@ -0,0 +1,555 @@
+            }
+        }
+
+        /// <summary>
+        /// Wss message handler
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        public override void OnMessage(object sender, WebSocketMessage e)
+        {
+            LastHeartbeatUtcTime = DateTime.UtcNow;
+        }
+
+        private void OnSocketMessage(object sender, WebSocketMessage e, Action<Messages.BaseMessage> handler)
+        {
+            try
+            {
+                if (_streamLocked)
+                {
+                    _messageBuffer.Enqueue(e);
+                    return;
+                }
+            }
+            catch (Exception err)
+            {
+                Log.Error(err);
+            }
+
+            OnMessageImpl(e, handler);
+        }
+
+        /// <summary>
+        /// Force reconnect
+        /// </summary>
+        protected override void Reconnect()
+        {
+            if (WebSocket.IsOpen)
+            {
+                WebSocket?.Close();
+            }
+            base.Reconnect();
+        }
+
+        private void ProcessSubscriptionRequest()
+        {
+            if (_subscriptionsPending) return;
+
+            _lastSubscribeRequestUtcTime = DateTime.UtcNow;
+            _subscriptionsPending = true;
+
+            Task.Run(async () =>
+            {
+                while (true)
+                {
+                    DateTime requestTime;
+                    List<Symbol> symbolsToSubscribe;
+                    lock (_lockerSubscriptions)
+                    {
+                        requestTime = _lastSubscribeRequestUtcTime.Add(_subscribeDelay);
+                        symbolsToSubscribe = SubscribedSymbols.ToList();
+                    }
+
+                    if (DateTime.UtcNow > requestTime)
+                    {
+                        // restart streaming session
+                        SubscribeSymbols(symbolsToSubscribe);
+
+                        lock (_lockerSubscriptions)
+                        {
+                            _lastSubscribeRequestUtcTime = DateTime.UtcNow;
+                            if (SubscribedSymbols.Count == symbolsToSubscribe.Count)
+                            {
+                                // no more subscriptions pending, task finished
+                                _subscriptionsPending = false;
+                                break;
+                            }
+                        }
+                    }
+
+                    await Task.Delay(200).ConfigureAwait(false);
+                }
+            });
+        }
+
+        private void SubscribeSymbols(List<Symbol> symbolsToSubscribe)
+        {
+            if (symbolsToSubscribe.Count == 0)
+                return;
+
+            //close current connection
+            if (TickerWebSocket.IsOpen)
+            {
+                TickerWebSocket.Close();
+            }
+            Wait(() => !TickerWebSocket.IsOpen);
```

Comment:
We could subscribe to `TickerWebSocket.Closed` + set an event that we watch here with a timeout? Same for `OnOpen`, so we don't pull at all 😃

Reply:
Check out https://github.com/QuantConnect/Lean/pull/4748/files#diff-048bd041ec2a67570c4245cbe4319c6bR209 😄

Reply:
Removed